lib/vanagon/component/rules.rb

```diff
@@ -0,0 +1,218 @@
+require 'vanagon/component'
+require 'vanagon/utilities/shell_utilities'
+require 'makefile'
+
+class Vanagon
+  class Component
+    # Vanagon::Component::Rules creates all Makefile rules for a given component.
+    class Rules
+
+      include Vanagon::Utilities::ShellUtilities
+
+      # Create methods that generate Makefile rules.
+      #
+      # This method cuts out some of the boilerplate of creating Makefile rules
+      # by creating methods and Makefile objects with a common name.
+      #
+      # @param target [Symbol] The rule target name.
+      # @param dependencies [Array<String>] An optional list of dependencies for the rule
+      # @yieldparam rule [Makefile::Rule] The generated Makefile rule
+      # @return [void]
+      #
+      # @!macro [attach] rule
+      #   @return [Makefile::Rule] The $1 rule
+      def self.rule(target, dependencies: [], &block)
+        define_method("#{target}_rule") do
+          Makefile::Rule.new("#{@component.name}-#{target}", dependencies: dependencies) do |rule|
+            instance_exec(rule, &block)
+          end
+        end
+      end
+
+      # @param component [Vanagon::Component] The component to create rules for.
+      # @param project [Vanagon::Project] The project associated with the component.
+      # @param platform [Vanagon::Platform] The platform where this component will be built.
+      def initialize(component, project, platform)
+        @component = component
+        @project = project
+        @platform = platform
+      end
+
+      # Generate all Makefile rules for this component.
+      #
+      # If the project has the cleanup attribute set, a cleanup rule will be included
+      # in the returned rules.
+      #
+      # @return [Array<Makefile::Rule>]
+      def rules
+        list = [
+          component_rule,
+          unpack_rule,
+          patch_rule,
+          configure_rule,
+          build_rule,
+          check_rule,
+          install_rule,
+          clean_rule,
+          clobber_rule,
+        ]
+        if @project.cleanup
+          list << cleanup_rule
+        end
+
+        list
+      end
+
+      # Generate a top level rule to build this component.
+      #
+      # @return [Makefile::Rule]
+      def component_rule
+        Makefile::Rule.new("#{@component.name}", dependencies: ["#{@component.name}-install"])
+      end
+
+      # Unpack the source for this component. The unpacking behavior depends on
+      # the source type of the component.
+      #
+      # @see [Vanagon::Component::Source]
+      rule("unpack", dependencies: ['file-list-before-build']) do |r|
+        if @component.extract_with
+          r.recipe << @component.extract_with
```

Comment:
I had a problem with needed to set environment variables during the unpack as well (mostly for windows). That change is in the vanagon_for_the_win branch, and I can add it in after, but it might be nice to add `@component.get_environment,` here as well